Logstash/logstash-input-azureblob/lib/logstash/inputs/azureblob.rb

```diff
@@ -117,15 +140,63 @@ def process(queue)
+
+          blob, header = @azure_blob.get_blob(@container, blob_name, {:end_range => (@file_head_bytes-1) }) if header.nil? unless @file_head_bytes.nil? or @file_head_bytes <= 0
+
+          if start_index == 0
+            # Skip the header since it is already read.
+            start_index = @file_head_bytes
+          else
+            # Adjust the offset when it is other than first time, then read till the end of the file, including the tail.
+            start_index = start_index - @file_tail_bytes
+            start_index = 0 if start_index < 0
+          end
+
+          blob, content = @azure_blob.get_blob(@container, blob_name, {:start_range => start_index} )
+
+          # content will be used to calculate the new offset. Create a new variable for processed content.
+          processed_content = content
+
+          is_json_codec = @codec.is_a? LogStash::Codecs::JSON
+          if (is_json_codec)
+            skip = processed_content.index '{'
+            processed_content = processed_content[skip..-1] unless skip.nil?
+          end #if
+
+          if is_json_codec && (@break_json_down_policy != 'do_not_break')
+            @logger.debug("codec is json and policy is not do_not_break")
+            
+            @break_json_batch_count = 1 if break_json_batch_count <= 0
+            tail = processed_content[-@file_tail_bytes..-1]
+            while (!processed_content.nil? && processed_content.length > @file_tail_bytes)
+              json_event, processed_content = get_jsons(processed_content, @break_json_batch_count)
+              @logger.debug("Got json: ========================")
+              @logger.debug("#{json_event[0..50]}...#{json_event[-50..-1]}")
+              @logger.debug("End got json: ========================")
+              @logger.debug("Processed content: #{processed_content[0..50]}...")
+              break if json_event.nil?
+              if @break_json_down_policy == 'with_head_tail'
+                @logger.debug("Adding json head/tails.")
+                json_event = "#{header}#{json_event}#{tail}"
+              end #if
+              @codec.decode(json_event) do |event|
+                decorate(event)
+                queue << event
+              end # decode
+            end
+          else
+            @logger.debug("Non-json codec or the policy is do not break")
+            # Putting header and content and tail together before pushing into event queue
+            processed_content = "#{header}#{processed_content}" unless header.nil? || header.length == 0
+            @codec.decode(processed_content) do |event|
+              decorate(event)
+              queue << event
+            end # decode
+          end #if
         ensure
           # Making sure the reader is removed from the registry even when there's exception.
           new_offset = start_index
+          new_offset = 0 if start_index == @file_head_bytes && content.nil? # Reset the offset when nothing has been read.
```

Comment:
I don't understand this. Why resetting the offset back to the beginning of the file, just because nothing new has been written by the writer?

Reply:
Yes, this is an edge case. If the start index is at the header while there is no content in the body, basically mean there's something wrong - breaking in the process / exception thrown / malformed json file and so on - which should not happen in normal case. I reset the index that the file will be treated as a non-touched file that it will be picked up and try again in the next iteration.

Reply:
ok, reasonable